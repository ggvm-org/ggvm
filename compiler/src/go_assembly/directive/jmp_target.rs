#[derive(Debug, PartialEq)]
pub(crate) enum JmpTarget {
    Addr(i64),
    Label(String),
}

macro_rules! impl_from_jmp_target {
    ($from_ty:ty => Label) => {
        impl From<$from_ty> for JmpTarget {
            fn from(v: $from_ty) -> Self {
                Self::Label(v.to_string())
            }
        }
    };
    ($from_ty:ty => Addr) => {
        impl From<$from_ty> for JmpTarget {
            fn from(v: $from_ty) -> Self {
                Self::Addr(i64::from(v))
            }
        }
    };
}

macro_rules! impl_from_jmp_target_label {
    ($($from_ty:ty),+) => {
        $(impl_from_jmp_target!($from_ty => Label);)+
    };
}

macro_rules! impl_from_jmp_target_addr {
    ($($from_ty:ty),+) => {
        $(impl_from_jmp_target!($from_ty => Addr);)+
    };
}

impl_from_jmp_target_label!(&str, String);
impl_from_jmp_target_addr!(u8, u16, u32, i8, i16, i32, i64);
