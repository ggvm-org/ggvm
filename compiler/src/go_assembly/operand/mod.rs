use std::fmt;

use super::{register::Register, register_with_offset::RegisterWithOffset};
#[derive(Debug, PartialEq)]
pub enum Operand {
    Ident(String),
    Int(i64),
    RegisterWithOffset(RegisterWithOffset),
}

impl fmt::Display for Operand {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Operand::Ident(s) => s.clone(),
            Operand::Int(n) => format!("${n}"),
            Operand::RegisterWithOffset(inner) => inner.to_string(),
        };
        write!(f, "{s}")
    }
}

macro_rules! impl_from_operand {
    ($from_ty:ty => Ident) => {
        impl From<$from_ty> for Operand {
            fn from(v: $from_ty) -> Self {
                Self::Ident(v.to_string())
            }
        }
    };
    ($from_ty:ty => RegisterWithOffset) => {
        impl From<$from_ty> for Operand {
            fn from(v: $from_ty) -> Self {
                Self::RegisterWithOffset(
                    crate::go_assembly::register_with_offset::RegisterWithOffset::from(v),
                )
            }
        }
    };
    ($from_ty:ty => $variant:ident) => {
        impl From<$from_ty> for Operand {
            fn from(v: $from_ty) -> Self {
                Self::$variant(v)
            }
        }
    };
}

impl_from_operand!(i64 => Int);
impl_from_operand!(&str => Ident);
impl_from_operand!(String => Ident);
impl_from_operand!(RegisterWithOffset => RegisterWithOffset);
impl_from_operand!(Register => RegisterWithOffset);

#[macro_export]
macro_rules! operand {
    ($offset:expr => $register_variant:ident) => {
        $crate::go_assembly::operand::Operand::RegisterWithOffset(crate::register_with_offset!($offset => $register_variant))
    };
    ($register:ident) => {
       $crate::go_assembly::operand::Operand::RegisterWithOffset(crate::register_with_offset!($register))
    };
    ($expr:expr) => {
        $crate::go_assembly::operand::Operand::from($expr)
    };
}

#[cfg(test)]
mod snapshots {
    use super::*;
    use insta::assert_display_snapshot;

    macro_rules! insta_test {
        ($testname:ident: $($testcases:expr),+) => {
            #[test]
            fn $testname() {
                $(assert_display_snapshot!($testcases);)+
            }
        };
    }

    insta_test!(operand: operand!(AX), operand!(16=>SP), operand!(1));
}
