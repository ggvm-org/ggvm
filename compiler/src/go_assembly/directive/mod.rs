use std::fmt;

use self::jmp_target::JmpTarget;

use super::operand::Operand;

#[macro_use]
pub mod jmp_target;

#[derive(Debug, PartialEq)]
pub(crate) enum Directive {
    Text { package: String, name: String },
    Subq(Operand, Operand),
    Leaq(Operand, Operand),
    Movq(Operand, Operand),
    Call { package: String, name: String },
    Addq(Operand, Operand),
    Ret,

    // CMPQ	SP, 16(R14)
    Cmpq(Operand, Operand),
    // PCDATA	$0, $-2
    PCData(Operand, Operand),

    // epi:
    Label(String),
    // NOP
    Nop,
    // JMP body
    Jmp(JmpTarget),
    // Jls	epi
    Jls(JmpTarget),
}

#[macro_export]
macro_rules! directive {
    (NOP) => {
        Directive::Nop
    };
    (RET) => {
        Directive::Ret
    };
    (CALL $package:ident.$name:ident) => {
        CALL!($package:ident.$name:ident)
    };
    (JMP $target:tt) => {
        JMP!($tt)
    };
    (@$label_name:ident) => {
        Directive::Label(stringify!($label_name))
    };
}

#[macro_export]
macro_rules! CALL {
    ($package:ident.$name:ident) => {
        Directive::Call {
            package: stringify!($package).to_string(),
            name: stringify!($name).to_string(),
        }
    };
}

macro_rules! define_jmp_macro {
    ($macro_name:ident, $variant:ident) => {
        #[macro_export]
        macro_rules! $macro_name {
            ($target:expr) => {
                Directive::$variant(JmpTarget::from($target))
            };
            (@$label:ident) => {
                Directive::$variant(JmpTarget::from(stringify!($label)))
            };
        }
    };
}

define_jmp_macro!(JMP, Jmp);
define_jmp_macro!(JLS, Jls);

impl fmt::Display for Directive {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Self::Text { package, name } => format!("TEXT	{}.{}(SB), 4, $0-0", package, name),
            Self::Subq(left, right) => format!("SUBQ	{}, {}", left, right),
            Self::Call { package, name } => format!("CALL    {package}Â·{name}(SB)"),
            Self::Addq(left, right) => format!("ADDQ	{}, {}", left, right),
            Self::Movq(left, right) => format!("MOVQ	{}, {}", left, right),
            Self::Cmpq(left, right) => format!("Cmpq	{}, {}", left, right),
            Self::Jls(target) => format!("Jls	{}", target),
            Self::PCData(left, right) => format!("PCDATA {}, {}", left, right),
            Self::Label(label_name) => format!("{}:", label_name),
            Self::Jmp(target) => format!("JMP	{}", target),
            Self::Nop => "NOP".to_string(),
            _ => unimplemented!(),
        };
        write!(f, "{s}")
    }
}

impl fmt::Display for JmpTarget {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Self::Addr(s) => s.to_string(),
            Self::Label(l) => l.to_string(),
        };
        write!(f, "{s}")
    }
}

#[cfg(test)]
mod snapshots {
    use super::*;
    use insta::assert_display_snapshot;

    macro_rules! insta_test {
        ($testname:ident: $($testcases:expr),+) => {
            #[test]
            fn $testname() {
                $(assert_display_snapshot!($testcases);)+
            }
        };
    }

    insta_test!(nop: directive!(NOP));
    insta_test!(call: CALL!(main.run));

    const TEST_JMP_TARGET_VAR: &str = "AAAAA";
    insta_test!(
        jmp: JMP!(33),
        JMP!("epi"),
        JMP!(TEST_JMP_TARGET_VAR),
        JMP!(@body)
    );

    insta_test!(
        jls: JLS!(33),
        JLS!("epi"),
        JLS!(TEST_JMP_TARGET_VAR),
        JLS!(@body)
    );
}
