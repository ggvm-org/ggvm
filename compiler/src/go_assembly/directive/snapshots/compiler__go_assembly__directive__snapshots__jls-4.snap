---
source: compiler/src/go_assembly/directive/mod.rs
assertion_line: 136
expression: JLS!(@ body)

---
Jls	body
